document current release steps and add step by step guides (#61)
mesosphere/kubeaddons-kommander: README.md

# Using and Releasing `kubeaddons-kommander` Repository

> This repository mostly follows the same rules that [`kubernetes-base-addons`](https://github.com/mesosphere/kubernetes-base-addons) does, please check documentation over there.

Reviewer: ```suggestion
This repository mostly follows the same rules that [`kubernetes-base-addons`](https://github.com/mesosphere/kubernetes-base-addons) does. Please familiarize yourself with that documentation, and then see the differences below.
```

- [RELEASE.md](https://github.com/mesosphere/kubernetes-base-addons/blob/master/RELEASE.md)
- [CONTRIBUTING.md](https://github.com/mesosphere/kubernetes-base-addons/blob/master/CONTRIBUTING.md)

## Deviations from `kubernetes-base-addons` release process

Kommander being a slightly different type of repository than `kubernetes-base-addons`, some details in releasing kommander are different.

- Pre-releases are only tagged with `testing-` tags, no `stable-` tags for pre-releases. The `-rc` that will acutally be the stable GA release will be _also_ tagged with `stable-` tag (and w/o `-rc` suffix)
- Kommander follows a more semver style versioning and is setup to be able to support multiple major and minor versions at the same time.

## Step by step guides to release Kommander versions (pre-releases, latest and previously released)

Only certain people are allowed to push to `testing` and `stable` branches. If you are asked to do a release, but dont have permissions to push, please ask your manager.

### Tag/release pre-releases and update update SOAK

1. fetch latest repo state: `git fetch`
1. checkout `testing` branch: `git checkout testing`

> 1. _merge_ `origin/master` into `testing` and push updated `testing` branch: `git merge origin/master && git push`

Reviewer 2: Are you italicizing merge for a reason?
```suggestion
1. merge `origin/master` into `testing` and push updated `testing` branch: `git merge origin/master && git push`
```

Author: yeah, rebasing or squashing would screw up github history in these cases, so its important to use `git merge`.

Reviewer 2: no i understand why you're using `git merge`, my question was why stress the word 'merge' in your instructions. Not sure it conveys "merge, not squash or rebase". Probably the way to make it most clear is not even explain but just list the commands to run as they seem clear enough:
```
1. git fetch
1. git checkout testing
1. git merge origin/master && git push
```

1. apply tags, kommander uses multiple tags at this point:
   1. apply and push "base" semver based tag: e.g. `git tag v1.1.0-beta.3 && git push origin v1.1.0-beta.3`
   1. apply and push "consumable" testing tag(s) for each supported k8s version: e.g. `git tag testing-1.16-1.1.0-beta.3 && git push origin testing-1.16-1.1.0-beta.3`
1. head to github and update release information for that prerelease: check [releases page](https://github.com/mesosphere/kubeaddons-kommander/releases) for up-to-date example. _Make sure to edit the "base" release_.

> 1. make sure SOAK gets updated, you can sync with people who tag `kubernetes-base-addons`.

Reviewer 2: Should we not include instructions to update soak? Isn't it updating the cluster.yaml and `konvoy deploy addons`?

Or is it because this is a public repo and we don't want to reference private ones?

Author: releasing kommander and updating soak are two different things, I'd think there is descriptions about how to update soak in soak repo 🤷

#### Hotfixing testing releases when problems are found on SOAK

> Instead of hotfixing testing released, we should create a new one.

Reviewer: ```suggestion
Instead of adding a hotfix to an already established release we should create a new one.
```

> When a major issue in "beta 3" is found on SOAK, we should follow the steps above to create "beta 4", and update SOAK with that. Same is true for RCs.

Reviewer 2: Interesting... 🤔 

Have you considered following KBAs approach to having a `testing` branch and whenever a fix is applied, soak can be updated by re-running `deploy addons` which will pull in new changes.

Author: we are following the kba approach here when it comes to the branch, but we also need to cut a new tag to ship to customers - its based on our discussion in this weeks leads meeting.

> In case we absolutely need to hotfix soak, we need to make sure to forwardport them to `master`:

Reviewer: ```suggestion
In case we absolutely need to hotfix soak, we need to make sure to push the fix to `master`:
```

Author: ```suggestion
In case we absolutely need to hotfix soak, we need to make sure to also push the fix to `master`:
```

ok @shaneutt ?

1. Create PR against `testing` branch _and_ forwardport the fix to `master` too.
1. Update SOAK.

### Tag/release latest stable version

Stable releases should be SOAKed for at least two weeks, after the soaking period was successful, they can be publicly released.

1. fetch latest repo state: `git fetch`
1. checkout `stable` branch: `git checkout stable`

> 1. _merge_ `origin/testing` into `stable` and push updated `stable` branch: `git merge origin/testing && git push`

Reviewer 2: ```suggestion
1. merge `origin/testing` into `stable` and push updated `stable` branch: `git merge origin/testing && git push`
```

1. apply and push "consumable" stable tag(s) for supported k8s versions: e.g. `git tag stable-1.16-1.1.0 && git push origin stable-1.16-1.1.0`
1. head to github and update release information for that prerelease: check [releases page](https://github.com/mesosphere/kubeaddons-kommander/releases) for up-to-date example. _Make sure to edit the "base" release_.

### Dealing with previously released stable versions

> _This needs get more details in how to deal with these situations in general._

Reviewer 2: Is this line necessary?

Author: probably not 🤷‍♂️ shall I remove it?

Reviewer 2: yes, doesn't seem like it adds any value

Author: ```suggestion

```
